### ISSUE TITLE
Bounded wildcard support?

### ISSUE THREAD

**Issue by tekacs** [lomigmegard/akka-http-cors#25]
This library uses `HttpOriginRange` from akka-http to perform its matching.

Unfortunately, there is no way to *extend* `HttpOriginRange` or `HttpOrigin` in order to support more sophisticated/alternate matching.

This means that whilst it's possible to match all hosts (`HttpOriginRange.*`) or a set of fixed hosts (`HttpOriginRange.Default(...)`), it's not possible for CORS purposes to match for example `*.somedomain.com`.

I would open this as an issue upstream, but it's unclear to me whether this is a request that's applicable to other uses of their `HttpOriginRange` or is specific to `akka-http-cors`'s use of it.

So in summary:

- Would it be possible here to add support for matching on bounded wildcards.
- Should this issue go upstream to `akka-http` instead?

**Comment by lomigmegard** [comment]
Hey, 
I knew one day someone would raise this issue ;)

Modifying `HttpOriginRange` or `HttpOrigin` is impracticable for two reasons:

- A `HttpOrigin` is strictly defined in [RFC-6454](https://tools.ietf.org/html/rfc6454#section-4) and it doesn't allow some *matching* segments. A `HttpOrigin` is used in many places and we can't modify his semantic for CORS.
- The range is used directly in the [`Access-Control-Allow-Origin` header](https://www.w3.org/TR/cors/#access-control-allow-origin-response-header) and the W3C defines its content strictly. 

I looked at other CORS implementations to see how they deal with this:

- The CORS filter from Dzhuvinov Software adds a new setting [`allowSubdomains`](http://software.dzhuvinov.com/cors-filter-configuration.html#cors.allowSubdomains).
    > If true the CORS filter will allow requests from any origin which is a subdomain origin of the allowed origins. A subdomain is matched by comparing its scheme and suffix (host name / IP address and optional port number). 
- The [CORS filter from eBay](https://github.com/eBay/cors-filter) doesn't allow it.
- [...] *please add more*

Adding a new setting `allowSubdomains` is possible, but this mechanism is not defined by the W3C. This setting would only change the matching logic, not the content of the `Access-Control-Allow-Origin` header.

Thoughts?

**Comment by tekacs** [comment]
> A HttpOrigin is strictly defined in RFC-6454 and it doesn't allow some matching segments. A HttpOrigin is used in many places and we can't modify his semantic for CORS.

So this is what I suspected and why I didn't raise this issue upstream with akka-http. I'm glad I posed the question here first; thanks for filling me in! :smile:

> Adding a new setting `allowSubdomains` is possible, but this mechanism is not defined by the W3C. This setting would only change the matching logic, not the content of the `Access-Control-Allow-Origin` header.

That would be perfect - the header does after all (as you say) expect the Origin passed to the server to be mirrored (as it is currently when you pass an HttpOriginRange on the server and the client receives only the domain that it sent).

I've tried to think about it some and I think that `allowSubdomains` would be flexible enough to cover all the use cases necessary for our use! Thanks again!

**Comment by aotd1** [comment]
If `allowSubdomains` would be global setting it will not cover the case where some origins should be checked strictly and some using wildcard:
```Scala
Seq("https://strict.com", "https://*.multidomain.site.com")
```

IMHO global `allowSubdomains` would be good, but weaker solution rather than a per-origin option. But I don't know for now how to implement it saving old behavior =)

**Comment by radusw** [comment]
Any plans to add this? `allowSubdomains` is really useful

**Comment by lomigmegard** [comment]
I created a PR for this feature, let me know what you think about it and it covers all your needs.
#46

**Comment by lomigmegard** [comment]
Released in https://github.com/lomigmegard/akka-http-cors/releases/tag/v04.0
